Add database plugin

LuckyPickleZZ opened the pull request:
Guys, I've added a database Plugin (Oceanbase) into dbeaver. Hope to contribute to the community.
This is my first Pull Request. If there are any deficiencies or other precautions, please tell me, and I'll make improvements.
<img width="500" alt="image" src="https://user-images.githubusercontent.com/36879124/128324100-d09a8834-3ccc-4af0-8cbe-dbf3dfeb81f0.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/36879124/128324119-ea21c0b5-70c1-499c-903f-17ca3c7a0359.png">
<img width="300" alt="image" src="https://user-images.githubusercontent.com/36879124/128324181-e157975f-ec64-461f-a354-5682c5260e35.png">

Commit:
Add OceanBase Plugin into Dbeaver

Commit:
Update configuration

Update configuration to introduce OceanBase plugin into product.

Commit:
Add tenant module

Override connectionPage to add tenant module into it. Override AuthDatabaseNative.

Commit:
Code format modification

Commit:
Change version

Commit:
Refactor project structure

Commit:
Add procedure/function to OB-MySQL mode

Commit:
Include dir 'icons' into properties

Commit:
Code format

Use eclipse-formatter-profile.xml to format code

Commit:
Remove test lib and icons

Commit:
Icons change

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,236 @@
+        Object readCost = nodeProps.get("COST");
+        if (readCost == null) {
+            readCost = nodeProps.get("COST");
+        }
```
Looks like a typo

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,236 @@
+        Object nodeName = nodeProps.get("table_name");
+        if (nodeName == null) {
+
+        } else {
```
Invert condition and remove empty branch

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,236 @@
+    public Number getNodeRowCount() {
+        Object rowCount = nodeProps.get("EST.ROWS");
+        if (rowCount == null) {
+            rowCount = nodeProps.get("EST.ROWS"); // MariaDB-specific plan
```
Looks like a typo

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,171 @@
+                                procedure == null ? null : JDBCUtils.escapeWildCards(session, procedure.getName()), "%")
+                        .getSourceStatement();
+            } else {
+                String queryFunctionString = "select * from mysql.proc where db='%s' and type='FUNCTION' and name='%s'";
+                return session
+                        .prepareStatement(String.format(queryFunctionString, owner.getName(), procedure.getName()));
```
Since you're already using prepared statements, please bind values instead of `String.format`-ting them

LuckyPickleZZ replied:
Thanks for review, I change it to statement.setString() method.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,171 @@
+        for (OceanbaseMySQLProcedure oceanbaseMySQLProcedure : oceanbaseProceduresCache.getAllObjects(monitor, this)) {
+            objects.add(oceanbaseMySQLProcedure);
+        }
+        return objects;
```
```suggestion
        return new ArrayList<>(oceanbaseProceduresCache.getAllObjects(monitor, this));
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,196 @@
+        if (name.equalsIgnoreCase("SHOW DB")) {
+            return DBUtils.findObject(getPrivileges(monitor), "Show databases", true);
+        }
+        ;
```
Unnecessary semicolon here

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,196 @@
+        for (OceanbaseMySQLCatalog oceanbaseMySQLCatalog : oceanbaseCatalogCache.getCachedObjects()) {
+            catalogs.add(oceanbaseMySQLCatalog);
+        }
+        return catalogs;
```
```suggestion
        return new ArrayList<>(oceanbaseCatalogCache.getCachedObjects());
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,74 @@
+    @Override
+    protected OceanbaseMySQLCatalog createDatabaseObject(DBRProgressMonitor monitor, DBECommandContext context,
+            Object container, Object copyFrom, Map<String, Object> options) throws DBException {
+        System.out.println("11111");
```
Please remove debug print

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public void renameObject(DBECommandContext commandContext, OceanbaseMySQLCatalog object,
+            Map<String, Object> options, String newName) throws DBException {
+        throw new DBException(
+                "Direct database rename is not yet implemented in MySQL. You should use export/import functions for that.");
+    }
```
If you don't support renaming objects, then don't implement `DBEObjectRenamer`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/plugin.xml:
```diff
@@ -0,0 +1,209 @@
+                  path="oceanbase">
+                <folder type="org.jkiss.dbeaver.ext.oceanbase.mysql.model.OceanbaseMySQLCatalog" label="%tree.databases.node.name" icon="#folder_schema" description="%tree.databases.node.tip">
+                    <items label="%tree.database.node.name" path="database" property="catalogs" icon="#database">
+                        <folder type="org.jkiss.dbeaver.ext.mysql.model.MySQLTable" label="%tree.tables.node.name" icon="#folder_table" description="%tree.tables.node.tip">
```
Is `MySQLTable` supposed to be there? Did you mean to use `OceanbaseTable` here, since you have such class?

LuckyPickleZZ replied:
Yes, I'm using OceanbaseTable instead. I missed it before

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public DBPDataSource getDataSource() {
+        // TODO Auto-generated method stub
```
Please remove TODO

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public DBCPlanStyle getPlanStyle() {
+        // TODO Auto-generated method stub
```
Please remove TODO

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,88 @@
+
+                rootNodes = nodes;
+            } catch (Exception e) {
+                // TODO: handle exception
```
Is it really important trying to handle exception here?

LuckyPickleZZ replied:
Sorry, I've changed it.

ShadelessFox reviewed (changes requested):
Thanks for contributing!

Commit:
Fixed issues

Fixed some codacy issues and requested changes.
https://app.codacy.com/gh/dbeaver/dbeaver/pullRequest?prid=7867479
https://github.com/dbeaver/dbeaver/pull/13449#pullrequestreview-723256362

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -29,28 +29,25 @@
-        try (JDBCPreparedStatement dbStat = session.prepareStatement(getPlanQueryString())) {
-            try (JDBCResultSet dbResult = dbStat.executeQuery()) {
-                List<OceanbasePlanNodeJSON> nodes = new ArrayList<>();
+        try{
+            JDBCPreparedStatement dbStat = session.prepareStatement(getPlanQueryString());
+            JDBCResultSet dbResult = dbStat.executeQuery();
```
The point of `try-with-resources` here was to close result set and query after exiting their scopes.

Please bring it back or call `.close()` on both `dbResult` and `dbStat` in `finally` block.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -66,9 +66,11 @@ protected JDBCStatement prepareChildrenStatement(JDBCSession session, OceanbaseM
-                return session
-                        .prepareStatement(String.format(queryFunctionString, owner.getName(), procedure.getName()));
+                String queryFunctionString = "select * from mysql.proc where db=? and type='FUNCTION' and name=?";
+                JDBCPreparedStatement statement = session.prepareStatement(String.format(queryFunctionString, owner.getName(), procedure.getName()));
```
Leftover and unnecessary `String.format`. Just inline `queryFunctionString` into `session.prepareStatement(...)`

LuckyPickleZZ replied:
Sorry, I missed this. I'll just inline `queryFunctionString`.

Commit:
Fixed issues

Fixed codacy issues and request changes.
https://app.codacy.com/gh/dbeaver/dbeaver/pullRequest?prid=7867479
https://github.com/dbeaver/dbeaver/pull/13449#pullrequestreview-723256362

ShadelessFox reviewed (approved):
👍🏻

LuckyPickleZZ commented:
@serge-rider @akilovich Could you please review this? Thanks a lot.

Commit:
Change plugin.xml

flambchap commented:
Hi @LuckyLeoZZ, thank you for the contribution. I'll review your PR today. Could you tell me please if it's possible to dump an OceanBase db using `mysqldump`? Thank you.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+package org.jkiss.dbeaver.ext.oceanbase.ui.internal;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLProcedure.java:
```diff
@@ -0,0 +1,34 @@
+package org.jkiss.dbeaver.ext.oceanbase.mysql.model;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,64 @@
+package org.jkiss.dbeaver.ext.oceanbase.mysql.model;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,168 @@
+package org.jkiss.dbeaver.ext.oceanbase.mysql.model;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,226 @@
+package org.jkiss.dbeaver.ext.oceanbase.model.plan;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+package org.jkiss.dbeaver.ext.oceanbase.model.plan;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+package org.jkiss.dbeaver.ext.oceanbase.model.plan;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/auth/OceanbaseAuthModelDatabaseNative.java:
```diff
@@ -0,0 +1,31 @@
+package org.jkiss.dbeaver.ext.oceanbase.model.auth;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/data/OceanbaseValueHandlerProvider.java:
```diff
@@ -0,0 +1,8 @@
+package org.jkiss.dbeaver.ext.oceanbase.data;
```
Please add a copyright notice. Also, could you tell me please why do we need this empty provider here?

LuckyPickleZZ replied:
This provider is empty for now. But in OracleMode, it'll be used, so I kept this provider for the time being.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+
+public class OceanbaseMessages extends NLS {
+    private static final String BUNDLE_NAME = "org.jkiss.dbeaver.ext.oceanbase.ui.internal.OceanbaseMessages"; //$NON-NLS-1$
+    public static String oceanbase_connection_page_label_port;
```
Please separate `BUNDLE_NAME` from localization variables with a blank line.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+    public static String oceanbase_connection_page_label_url;
+    public static String oceanbase_connection_page_label_connection;
+    public static String oceanbase_connection_page_label_tenant;
+    static {
```
Please separate localization variables from the static initializer with a blank line.

LuckyPickleZZ commented:
> Hi @LuckyLeoZZ, thank you for the contribution. I'll review your PR today. Could you tell me please if it's possible to dump an OceanBase db using `mysqldump`? Thank you.

Thanks a lot! And yes, Oceanbase is compatible with MySQL protocol, so `mysqldump` can be used to back up the data in OB. When you need to restore these data, you can do it by executing these `INSERT` statements.

LuckyPickleZZ commented:
If you want to try OceanBase, here is a deployment guide for Community Edition. 
1. Install OceanBase Deploy 
`sudo yum install -y yum-utils`
 `sudo yum-config-manager --add-repo https://mirrors.aliyun.com/oceanbase/OceanBase.repo`
 `sudo yum install -y ob-deploy`
 `source /etc/profile.d/obd.sh`
2. Download [https://github.com/oceanbase/obdeploy/blob/master/example/autodeploy/single-example.yaml](url)
 Modify the `servers` IP to your own server IP
3. Execute `obd cluster autodeploy [NAME] -c single-example.yaml -A` to deploy.
4. Execute `obd cluster start [NAME]` to start server.
5. Connect OceanBase Community Edition by `mysql -h [HOST] -u root -P 2881`
6. OBD has multi-level commands. You can use the `-H / -help` option in each level to view the help information of the subcommand.

Commit:
Fix OceanBaseView 'column' information not shown

Use 'desc SCHEMA.VIEW' command to get column information.
Display column info by OceanbaseMySQLViewColumn.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/OSGI-INF/l10n/bundle.properties:
```diff
@@ -0,0 +1,212 @@
+#Generated by ResourceBundle Editor (http://essiembre.github.io/eclipse-rbe/)
```
This file is unnecessary since we already have one in the MySQL bundle.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/pom.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please use 4 whitespaces for indentation.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/OceanbaseDataSourceProvider.java:
```diff
@@ -0,0 +1,47 @@
+public class OceanbaseDataSourceProvider extends MySQLDataSourceProvider {
+
+    public OceanbaseDataSourceProvider() {
+    }
```
Please delete this redundant default constructor.

Commit:
Fix codacy issues

https://app.codacy.com/gh/dbeaver/dbeaver/pullRequest?prid=7867479

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/auth/OceanbaseAuthModelDatabaseNative.java:
```diff
@@ -0,0 +1,31 @@
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+import org.jkiss.utils.CommonUtils;
+
+public class OceanbaseAuthModelDatabaseNative extends AuthModelDatabaseNative {
```
Please make add `AuthModelDatabaseNativeCredentials` as a generic parameter.

LuckyLeoZZ replied:
Sorry, I don't understand how this can be fixed.. Could you please explain it further?

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/auth/OceanbaseAuthModelDatabaseNative.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public Object initAuthentication(@NotNull DBRProgressMonitor monitor, @NotNull DBPDataSource dataSource,
+            AuthModelDatabaseNativeCredentials credentials, DBPConnectionConfiguration configuration,
```
The absence of `NotNull` annotations on this line makes the reader (or just me, idk) think that those arguments can be `null`, but in fact, they can't.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+import com.google.gson.JsonPrimitive;
+
+public class OceanbasePlanAnalyzer extends AbstractExecutionPlanSerializer implements DBCQueryPlanner {
+    private OceanbaseMySQLDataSource dataSource;
```
Field 'dataSource' may be 'final'.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+        this.dataSource = dataSource;
+    }
+
+    public OceanbasePlanJSON explain(JDBCSession session, String query) throws DBCException {
```
This method can be `private static`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    public OceanbasePlanJSON explain(JDBCSession session, String query) throws DBCException {
+        final SQLDialect dialect = SQLUtils.getDialectFromObject(session.getDataSource());
+        final String plainQuery = SQLUtils.stripComments(dialect, query).toUpperCase();
+        final String firstKeyword = SQLUtils.getFirstKeyword(dialect, plainQuery);
+        if (!"SELECT".equalsIgnoreCase(firstKeyword) && !"WITH".equalsIgnoreCase(firstKeyword)) {
+            throw new DBCException("Only SELECT statements could produce execution plan");
+        }
```
Please address this code duplication (the identical code can be found in `MySQLOlanAnalizer`).

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    @Override
+    public void serialize(Writer planData, DBCPlan plan) throws IOException, InvocationTargetException {
```
This method never throws `InvocationTargetException`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+
+            @Override
+            public String version() {
+                return "json";
```
There is a corresponding constant in `MySQLConstants`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    @Override
+    public DBCPlan deserialize(Reader planData) throws IOException, InvocationTargetException {
```
This method never throws `IOException`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanAnalyzer.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    public DBCPlan planQueryExecution(DBCSession session, String query, DBCQueryPlannerConfiguration configuration)
+            throws DBException {
```
The exception's type can be narrowed to `DBCException`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+
+public class OceanbasePlanJSON extends AbstractExecutionPlan {
+    protected OceanbaseMySQLDataSource dataSource;
+    protected String query;
```
These fields can be private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/OSGI-INF/l10n/bundle_zh.properties:
```diff
@@ -0,0 +1,11 @@
+#Generated by ResourceBundle Editor (http://essiembre.github.io/eclipse-rbe/)
```
The same is true for this file.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+    public static String oceanbase_connection_page_label_port;
+    public static String oceanbase_connection_page_label_host;
+    public static String oceanbase_connection_page_label_database;
+    public static String oceanbase_connection_page_label_password;
```
This field is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+    public static String oceanbase_connection_page_label_host;
+    public static String oceanbase_connection_page_label_database;
+    public static String oceanbase_connection_page_label_password;
+    public static String oceanbase_connection_page_label_user;
```
This field is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.java:
```diff
@@ -0,0 +1,22 @@
+    public static String oceanbase_connection_page_label_password;
+    public static String oceanbase_connection_page_label_user;
+    public static String oceanbase_connection_page_label_url;
+    public static String oceanbase_connection_page_label_connection;
```
This field is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.properties:
```diff
@@ -0,0 +1,8 @@
+oceanbase_connection_page_label_port=Port
+oceanbase_connection_page_label_host=Host
+oceanbase_connection_page_label_database=Database
+oceanbase_connection_page_label_password=Password
```
This message is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.properties:
```diff
@@ -0,0 +1,8 @@
+oceanbase_connection_page_label_database=Database
+oceanbase_connection_page_label_password=Password
+oceanbase_connection_page_label_url=JDBC URL
+oceanbase_connection_page_label_user=User
```
This message is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/OceanbaseMessages.properties:
```diff
@@ -0,0 +1,8 @@
+oceanbase_connection_page_label_password=Password
+oceanbase_connection_page_label_url=JDBC URL
+oceanbase_connection_page_label_user=User
+oceanbase_connection_page_label_connection=Connection
```
This message is not used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseTable.java:
```diff
@@ -0,0 +1,37 @@
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+import org.jkiss.dbeaver.model.struct.DBSEntity;
+
+public class OceanbaseTable extends MySQLTable implements DBPObjectStatistics {
```
I'm not convinced that we need this class.

LuckyPickleZZ replied:
Sorry, I'll remove this class. I missed this.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/Activator.java:
```diff
@@ -0,0 +1,84 @@
+public class Activator extends AbstractUIPlugin {
+
+    // The plug-in ID
+    public static final String PLUGIN_ID = "org.jkiss.dbeaver.ext.oceanbase";
```
Can be private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/Activator.java:
```diff
@@ -0,0 +1,84 @@
+     * The constructor
+     */
+    public Activator() {
+    }
```
Can be safely deleted.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/internal/Activator.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    public static Activator getDefault() {
+        return plugin;
+    }
```
Can be safely deleted.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLView.java:
```diff
@@ -0,0 +1,64 @@
+
+    public OceanbaseMySQLView(MySQLCatalog catalog) {
+        super(catalog);
+    }
```
Please remove the unused constructor.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLView.java:
```diff
@@ -0,0 +1,64 @@
+        super(catalog);
+    }
+
+    public OceanbaseMySQLView(MySQLCatalog catalog, ResultSet dbResult) {
```
Access can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLView.java:
```diff
@@ -0,0 +1,64 @@
+        return ((OceanbaseMySQLCatalog) getContainer()).oceanbaseTableCache.getChild(monitor,
+                (OceanbaseMySQLCatalog) getContainer(), this, attributeName);
+    }
+
```
In this class several times you call `getContainer()` and cast the result on the spot. I believe it's cleaner to override `getContainer()` just like `RedshiftExternalTable `does.

LuckyPickleZZ replied:
Thanks, I've overrided this method.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLView.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * OceanbaseView
+ */
+public class OceanbaseMySQLView extends MySQLView {
```
Please rename the class to `OceanbaseView`.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLViewColumn.java:
```diff
@@ -0,0 +1,65 @@
+package org.jkiss.dbeaver.ext.oceanbase.mysql.model;
```
Please add a copyright notice.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLProcedure.java:
```diff
@@ -0,0 +1,34 @@
+public class OceanbaseMySQLProcedure extends MySQLProcedure {
+    private OceanbaseMySQLCatalog container = (OceanbaseMySQLCatalog) getContainer();
+
+    public OceanbaseMySQLProcedure(MySQLCatalog catalog) {
```
Can be safely deleted.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLProcedure.java:
```diff
@@ -0,0 +1,34 @@
+        super(catalog);
+    }
+
+    public OceanbaseMySQLProcedure(MySQLCatalog catalog, ResultSet dbResult) {
```
Can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLViewColumn.java:
```diff
@@ -0,0 +1,65 @@
+import org.jkiss.utils.CommonUtils;
+
+public class OceanbaseMySQLViewColumn extends MySQLTableColumn {
+    public OceanbaseMySQLViewColumn(MySQLTableBase table, ResultSet dbResult) throws DBException {
```
Can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLViewColumn.java:
```diff
@@ -0,0 +1,65 @@
+        if (defaultValue != null) {
+            DBPDataKind dataKind = getDataKind();
+
+            if (dataKind == DBPDataKind.STRING && !SQLConstants.NULL_VALUE.equals(defaultValue)
+                    && !SQLUtils.isStringQuoted(getDataSource(), defaultValue)) {
+                defaultValue = SQLUtils.quoteString(getDataSource(), defaultValue);
+            } else if (dataKind == DBPDataKind.DATETIME && !defaultValue.isEmpty()
+                    && Character.isDigit(defaultValue.charAt(0))) {
+                defaultValue = "'" + defaultValue + "'";
+            } else {
+                setDefaultValue(defaultValue);
+            }
```
`defaultValue` is assigned in the first two branches, but never used. In general, looks like some kind of bug hides here.

LuckyPickleZZ replied:
Yes, you reviewed it very carefully. I really wrote the wrong logic in this place. The `else` branch should be removed.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,64 @@
+        );
+    }
+
+    @Override
+    protected void addObjectDeleteActions(DBRProgressMonitor monitor, DBCExecutionContext executionContext,
+            List<DBEPersistAction> actions,
+            SQLObjectEditor<OceanbaseMySQLCatalog, OceanbaseMySQLDataSource>.ObjectDeleteCommand command,
+            Map<String, Object> options) throws DBException {
+        actions.add(new SQLDatabasePersistAction("Drop schema", "DROP SCHEMA `" + command.getObject().getName() + "`"));
+    }
```
It looks like the content of this and the overridden method are identical.

LuckyPickleZZ replied:
Sorry, I guess you read this place wrong. This class `OceanbaseMySQLDatabaseManager` is extended from `SQLObjectEditor`, in which this method is `abstract` type.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,64 @@
+        return new OceanbaseMySQLCatalog((OceanbaseMySQLDataSource) container, null);
+    }
+
+    @Override
+    protected void addObjectCreateActions(DBRProgressMonitor monitor, DBCExecutionContext executionContext,
+            List<DBEPersistAction> actions,
+            SQLObjectEditor<OceanbaseMySQLCatalog, OceanbaseMySQLDataSource>.ObjectCreateCommand command,
+            Map<String, Object> options) throws DBException {
+        final OceanbaseMySQLCatalog catalog = command.getObject();
+        final StringBuilder script = new StringBuilder("CREATE SCHEMA `" + catalog.getName() + "`");
+        appendDatabaseModifiers(catalog, script);
+        actions.add(new SQLDatabasePersistAction("Create schema", script.toString()) // $NON-NLS-2$
+        );
+    }
```
It looks like the content of this and the overridden method are identical.

LuckyPickleZZ replied:
Same as the previous one。

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDatabaseManager.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    protected OceanbaseMySQLCatalog createDatabaseObject(DBRProgressMonitor monitor, DBECommandContext context,
+            Object container, Object copyFrom, Map<String, Object> options) throws DBException {
```
This method never throws exceptions.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,190 @@
+        return dataTypeCache.getCachedObject(typeID);
+    }
+
+    public boolean isMySQLMode() {
+        return tenantType.equals("MySQL");
+    }
```
This method is unused.

LuckyPickleZZ replied:
This method will be used in `plugin.xml`, so I choose to keep it for now.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,190 @@
+            return null;
+        }
+        return catalog.getTable(monitor, tableName);
+    }
```
This method is also unused.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public Class<? extends MySQLCatalog> getPrimaryChildType(@Nullable DBRProgressMonitor monitor) throws DBException {
+        return OceanbaseMySQLCatalog.class;
+    }
```
These methods never throw exceptions.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,190 @@
+        return OceanbaseMySQLCatalog.class;
+    }
+
+    public OceanbaseCatalogCache getOceanbaseCatalogCache() {
```
This method can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLDataSource.java:
```diff
@@ -0,0 +1,190 @@
+        @Override
+        protected OceanbaseMySQLCatalog fetchObject(@NotNull JDBCSession session,
+                @NotNull OceanbaseMySQLDataSource owner, @NotNull JDBCResultSet resultSet)
+                throws SQLException, DBException {
```
These exceptions are never thrown.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+public class OceanbaseMySQLCatalog extends MySQLCatalog {
+    private final OceanbaseMySQLDataSource dataSource;
+    final OceanbaseProceduresCache oceanbaseProceduresCache = new OceanbaseProceduresCache();
+    final OceanbaseTableCache oceanbaseTableCache = new OceanbaseTableCache();
```
Please make these fields private and provide getters for them instead.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+    public static class OceanbaseTableCache
+            extends JDBCStructLookupCache<OceanbaseMySQLCatalog, MySQLTableBase, MySQLTableColumn> {
+
+        OceanbaseTableCache() {
```
Can be private.

LuckyPickleZZ replied:
Change it to package private instead of private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+    final OceanbaseProceduresCache oceanbaseProceduresCache = new OceanbaseProceduresCache();
+    final OceanbaseTableCache oceanbaseTableCache = new OceanbaseTableCache();
+
+    public static class OceanbaseTableCache
```
Please move this class to the bottom of the outer class declaration.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+        @Override
+        protected MySQLTableBase fetchObject(@NotNull JDBCSession session, @NotNull OceanbaseMySQLCatalog owner,
+                @NotNull JDBCResultSet dbResult) throws SQLException, DBException {
```
These exceptions are never thrown.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+        @Override
+        protected MySQLTableColumn fetchChild(@NotNull JDBCSession session, @NotNull OceanbaseMySQLCatalog owner,
+                @NotNull MySQLTableBase table, @NotNull JDBCResultSet dbResult) throws SQLException, DBException {
```
These exceptions are never thrown.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+    }
+
+    static class OceanbaseProceduresCache
```
Please move this class to the bottom of the outer class declaration.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+    static class OceanbaseProceduresCache
+            extends JDBCStructLookupCache<OceanbaseMySQLCatalog, OceanbaseMySQLProcedure, MySQLProcedureParameter> {
+
+        public OceanbaseProceduresCache() {
```
Can be private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+            if (procedure.getProcedureType().equals(DBSProcedureType.PROCEDURE)) {
+                return session.getMetaData()
+                        .getProcedureColumns(owner.getName(), null,
+                                procedure == null ? null : JDBCUtils.escapeWildCards(session, procedure.getName()), "%")
```
Null check is redundant

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+                return new MySQLProcedureParameter(parent, columnName, typeName, valueType, position, columnSize, scale,
+                        precision, notNull, parameterType);
+            } else {
+                String[] paramList = JDBCUtils.safeGetString(dbResult, "returns").split("\\(");
```
`safeGetString` can return `null`. Potential `NPE`

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+        @Override
+        protected OceanbaseMySQLProcedure fetchObject(JDBCSession session, OceanbaseMySQLCatalog owner,
+                JDBCResultSet resultSet) throws SQLException, DBException {
```
Exceptions are never thrown.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+    }
+
+    public OceanbaseMySQLCatalog(OceanbaseMySQLDataSource dataSource, ResultSet dbResult) {
```
Can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+
+    public OceanbaseProceduresCache getOceanbaseProceduresCache() {
+        return this.oceanbaseProceduresCache;
+    }
```
Never used.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/mysql/model/OceanbaseMySQLCatalog.java:
```diff
@@ -0,0 +1,335 @@
+    @NotNull
+    @Override
+    public MySQLDataSource getDataSource() {
+        return (OceanbaseMySQLDataSource) dataSource;
```
Redundant cast.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,226 @@
+    private OceanbasePlanNodeJSON parent;
+    private String name;
+    private JsonObject object;
+    private Map<String, String> nodeProps = new LinkedHashMap<>();
```
Can be final.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,226 @@
+
+public class OceanbasePlanNodeJSON extends AbstractExecutionPlanNode implements DBCPlanCostNode, DBPPropertySource {
+
+    private OceanbasePlanNodeJSON parent;
```
Can be private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,226 @@
+    private Map<String, String> nodeProps = new LinkedHashMap<>();
+    private List<OceanbasePlanNodeJSON> nested = new ArrayList<>();
+
+    public OceanbasePlanNodeJSON(OceanbasePlanNodeJSON parent, String name, JsonObject object) {
+        this.parent = parent;
+        this.name = name;
+        this.object = object;
+
+        parseObject(name, object);
+    }
+
+    public OceanbasePlanNodeJSON(OceanbasePlanNodeJSON parent, Map<String, String> attributes) {
+        this.parent = parent;
+        this.nodeProps.putAll(attributes);
+    }
+
+    public Map<String, String> getNodeProps() {
+        return nodeProps;
+    }
```
Can be package-private.

It's also worth wrapping `nodeProps` in `Collections.unmodifiableMap()` for encapsulation reasons.

LuckyLeoZZ replied:
If I wrap `nodeProps` into `unmodifiableMap()`, it can't be modified. But I still need to call its `put` method, so I maybe keep it.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanNodeJSON.java:
```diff
@@ -0,0 +1,226 @@
+        return nested;
+    }
+
+    public Object getProperty(String name) {
```
Can be package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+    protected OceanbaseMySQLDataSource dataSource;
+    protected String query;
+
+    private static final Gson gson = new Gson();
```
Please move this field to the very top of the class declaration.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+
+    private static final Gson gson = new Gson();
+
+    private List<OceanbasePlanNodeJSON> rootNodes;
```
Can be final.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+
+    private List<OceanbasePlanNodeJSON> rootNodes;
+
+    public OceanbasePlanJSON(JDBCSession session, String query) throws DBCException {
```
Can package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public OceanbasePlanJSON(MySQLDataSource dataSource, String query, List<OceanbasePlanNodeJSON> rootNodes) {
```
Can package-private.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/model/plan/OceanbasePlanJSON.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public List<OceanbasePlanNodeJSON> getPlanNodes(Map<String, Object> options) {
+        return rootNodes;
```
It's worth wrapping `rootNodes` in `Collections.unmodifiableList()` for encapsulation reasons.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/views/OceanbaseConnectionPage.java:
```diff
@@ -0,0 +1,202 @@
+    
+    private static ImageDescriptor logoImage = Activator.getImageDescriptor("icons/ob_logo.png");
+
+    public OceanbaseConnectionPage() {
+    }
```
This constructor is redundant.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/views/OceanbaseConnectionPage.java:
```diff
@@ -0,0 +1,202 @@
+    private Text databaseText;
+    private Text tenantText;
+    
+    private static ImageDescriptor logoImage = Activator.getImageDescriptor("icons/ob_logo.png");
```
Please move this field to the very top of the class and make it final.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/views/OceanbaseConnectionPage.java:
```diff
@@ -0,0 +1,202 @@
+    public OceanbaseConnectionPage() {
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+    }
```
This method is redundant.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/views/OceanbaseConnectionPage.java:
```diff
@@ -0,0 +1,202 @@
+        addrGroup.setLayoutData(gd);
+
+        {
+            Group hostGroup = UIUtils.createControlGroup(addrGroup, "Connection", 4, GridData.FILL_HORIZONTAL, 0);
```
Please add internationalization.

flambchap commented on plugins/org.jkiss.dbeaver.ext.oceanbase/src/org/jkiss/dbeaver/ext/oceanbase/ui/views/OceanbaseConnectionPage.java:
```diff
@@ -0,0 +1,202 @@
+                try {
+                    saveSettings(site.getActiveDataSource());
+                } catch (Exception e) {
+                    setMessage(e.getMessage());
```
Please log this exception. It can created like this:
```Java
private static final Log log = Log.getLog(OceanbaseConnectionPage.class);
```
Place this statement to the very top of the class, even above every other `static final` variables.

flambchap reviewed (commented):
Sorry that it took me so long to review your PR. I want to say that you made a fantastic job, and I generally really like your code style. There are 3 major points where your work can be improved:

1. Core and UI packages currently live in one bundle. I want to ask you to separate them into two separate bundles: `oceanbase` and `oceanbase.ui`.
2. I also want to ask you to move all the classes from `ext.oceanbase.mysql.model` to `ext.oceanbase.model`. It would also be fantastic if you'll remove `MySQL` from the names of the classes from this package, e.g. `OceanbaseMySQLView` --> `OceanbaseView`.
3. There is a potential NPE in one of the methods, please find it in one of the comments below.

The other ones are quite minor.

I also want to tell you that I'm going on vacation tomorrow and won't be able to review your changes for a while, but I'm convinced you'll make a good job. Thank you again!

Commit:
Fix requested changes by Akilovich

https://github.com/dbeaver/dbeaver/pull/13449#pullrequestreview-726334149

Commit:
Separate bundle `oceanbase` to `oceanbase` and `oceanbase.ui`

LuckyPickleZZ commented:
> Sorry that it took me so long to review your PR. I want to say that you made a fantastic job, and I generally really like your code style. There are 3 major points where your work can be improved:
> 
> 1. Core and UI packages currently live in one bundle. I want to ask you to separate them into two separate bundles: `oceanbase` and `oceanbase.ui`.
> 2. I also want to ask you to move all the classes from `ext.oceanbase.mysql.model` to `ext.oceanbase.model`. It would also be fantastic if you'll remove `MySQL` from the names of the classes from this package, e.g. `OceanbaseMySQLView` --> `OceanbaseView`.
> 3. There is a potential NPE in one of the methods, please find it in one of the comments below.
> 
> The other ones are quite minor.
> 
> I also want to tell you that I'm going on vacation tomorrow and won't be able to review your changes for a while, but I'm convinced you'll make a good job. Thank you again!

Thank you very much for your careful review, and I've fixed these minor problems. For the above three suggestions, here are my replies.
1. I've separated the bundle into to plugins: `oceanbase` and `oceanbase.ui`.
2. For business reasons, OceanBase is compatible with MySQL and Oracle protocol, that is, there are two unrelated operation modes. In the current PR, only MySQL schema is involved, so the format of packages and classes has such names. Later, I will open a PR for OceanBase-Oracle mode. In Oracle mode, the naming format will be named in the format of ext.oceanbase.oracle.model/OceanbaseOracleView. These two modes are in one bundle, so I have to name Java packages and classes this way.
3. I added a judgment to avoid that NPE.

Thanks again~ If there is any problem, please let me know.

Commit:
Modify debug/DBeaver.product.launch

Add bundle oceanbase.ui

serge-rider commented:
@LuckyLeoZZ Thanks!
We'll merge it into devel and start testing it to include Oceanbase support in DBeaver CE 21.2.

serge-rider commented:
@LuckyLeoZZ let's continue discussion in #13596
